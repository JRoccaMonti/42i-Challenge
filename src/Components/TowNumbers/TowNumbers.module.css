.towNumbersContainer{
    display: flex;
    flex-direction: column;
    margin: 10px;
}

.inputNumbersContainer{
    margin: 10px;
    display: flex;
}

.inputNumbers{
    margin-left: 10px;
}
.resultContainer{
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-direction: row;
    margin: 10px;
}

.targetNumbersContainer{
    margin: 10px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-direction: row;
    
}
.resultContainer p{
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-direction: row;
    margin: 0;
}

.errorsContainer{
    display: flex;
    flex-direction: column;
    margin: 10px;
    color: red;
}

.errorsContainer p{
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-direction: row;
    margin: 0;
}