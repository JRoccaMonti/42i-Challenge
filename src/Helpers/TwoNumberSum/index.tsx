export const twoNumberSum = (input: number[], target: number): number[] => {
  const seen : { [key: number]: number } = {}
  for (let i = 0; i < input.length; i++) {
    const complement = target - input[i];
    if (Object.prototype.hasOwnProperty.call(seen , complement)) {
      return [complement,input[i]];
    }
    seen [input[i]]=i;
  }
  
  return [];
};


