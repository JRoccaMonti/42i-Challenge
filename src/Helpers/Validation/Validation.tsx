export const validateArrayNumbers = (
    numbers: string, 
    validateDuplicates: boolean = false,
    AllowNegatives: boolean = true,
  ): string[] => {
    const errors: string[] = [];
    
    if (numbers.startsWith(',') || numbers.endsWith(',')) {
      errors.push('La lista de números no debe comenzar ni terminar con una coma.');
    }
    if (AllowNegatives) {
      if (/[^0-9-],|,[^0-9-]/.test(numbers)) {
        errors.push('Las comas deben estar separadas por números enteros.');
      }
    }else{
      if (/[^0-9],|,[^0-9]/.test(numbers)) {
        errors.push('Las comas deben estar separadas por números enteros positivos.');
      }
    }
  
    const numArray = numbers.split(',')
      .map(num => num.trim())
      .filter(num => num !== '');
  
    numArray.forEach(num => {
      const parsedNum = parseFloat(num);
      if (isNaN(parsedNum)) {
        errors.push(`El valor "${num}" no es un número válido.`);
      }
    });
  
    if (validateDuplicates) {
      const numSet = new Set(numArray);
      if (numSet.size !== numArray.length) {
        errors.push('La lista de números no debe contener duplicados.');
      }
    }
  
    return errors;
};  
  
export const validateTarget = (target: number | ''): string[] => {
    const errors: string[] = [];
    if (target === '' || !Number.isInteger(target)) {
        errors.push('Por favor, ingresa un objetivo válido (número entero).');
    }
   return errors;
};
  
export const validateNumArray = (numArray: number[]): string[] => {
    const errors: string[] = [];
    if (numArray.length === 0) {
        errors.push('Por favor, ingresa al menos un número entero.');
    }
    return errors;
};
  