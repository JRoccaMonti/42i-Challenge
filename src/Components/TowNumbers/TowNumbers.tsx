import React, { useState } from 'react';
import TwoNumberSum from '../../Helpers/TwoNumberSum';
import { validateArrayNumbers, validateTarget, validateNumArray } from '../../Helpers/Validation/Validation';

const TowNumbers: React.FC = () => {
  const [numbers, setNumbers] = useState<string>('');
  const [target, setTarget] = useState<number | ''>('');
  const [result, setResult] = useState<number[]>([]);
  const [errors, setErrors] = useState<string[]>([]);
  
  const validateInputs = () => {
    const newErrors: string[] = [];
    newErrors.push(...validateArrayNumbers(numbers,true));
    const numArray = numbers.split(',')
      .map(num => parseFloat(num.trim()))
      .filter(num => !isNaN(num));
    newErrors.push(...validateNumArray(numArray));
    newErrors.push(...validateTarget(target));
    setErrors(newErrors);
    return newErrors.length === 0;
  };
  const handleCalculate = () => {
    if (!validateInputs()) return;
    const numArray = numbers.split(',').map(num => parseFloat(num.trim())).filter(num => !isNaN(num));
    if (target !== '' && numArray.length > 0) {
      const sumResult = TwoNumberSum(numArray, target);
      setResult(sumResult);
    }
  };

  return (
    <div>
      <div>
        <label>
          NÃºmeros (separados por comas):
          <input
            type="text"
            value={numbers}
            onChange={(e) => setNumbers(e.target.value)}
            placeholder="Ej: 1, 2, 3, 4"
          />
        </label>
      </div>
      <div>
        <label>
          Objetivo:
          <input
            type="number"
            value={target}
            onChange={(e) => setTarget(Number(e.target.value))}
            placeholder="Ej: 5"
          />
        </label>
      </div>
      <button onClick={handleCalculate}>Calcular</button>
      {errors.length > 0 && (
        <div style={{ color: 'red' }}>
          {errors.map((error, index) => (
            <p key={index}>{error}</p>
          ))}
        </div>
      )}
      <div>
        <p>Resultado: [ {result.length > 0 ? result.join(', ') : ''} ]</p>
      </div>
    </div>
  );
};

export default TowNumbers;
