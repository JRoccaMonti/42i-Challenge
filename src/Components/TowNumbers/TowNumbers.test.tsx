import { render, fireEvent, screen  } from '@testing-library/react';
import '@testing-library/jest-dom'
import TowNumbers from "./TowNumbers";
import TwoNumberSum from "../../Helpers/TwoNumberSum";
import { validateArrayNumbers, validateNumArray, validateTarget } from '../../Helpers/Validation/Validation';

jest.mock('../../Helpers/TwoNumberSum');
jest.mock('../../Helpers/Validation/Validation');

describe('TowNumbers Component', () => {
  it('Debería coincidir con la snapshot', () => {
    const { asFragment } = render(<TowNumbers />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Debería mostrar errores de validación para entradas no válidas', () => {
    (validateArrayNumbers as jest.Mock).mockReturnValue(['Error en la validación de array']);
    (validateTarget as jest.Mock).mockReturnValue(['Error en la validación del objetivo']);
    (validateNumArray as jest.Mock).mockReturnValue([]);

    render(<TowNumbers />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: 'abc, def' } });
    fireEvent.change(screen.getByPlaceholderText('Ej: 5'), { target: { value: 'a' } });
    fireEvent.click(screen.getByText('Calcular'));

    expect(screen.getByText('Error en la validación de array')).toBeInTheDocument();
    expect(screen.getByText('Error en la validación del objetivo')).toBeInTheDocument();
  });

  it('No debería calcular el resultado si hay errores de validación', () => {
    (validateArrayNumbers as jest.Mock).mockReturnValue(['Error en la validación de array']);
    (validateTarget as jest.Mock).mockReturnValue(['Error en la validación del objetivo']);
    (validateNumArray as jest.Mock).mockReturnValue([]);

    render(<TowNumbers />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: 'abc, def' } });
    fireEvent.change(screen.getByPlaceholderText('Ej: 5'), { target: { value: 'a' } });
    fireEvent.click(screen.getByText('Calcular'));

    expect(TwoNumberSum).not.toHaveBeenCalled();
    expect(screen.getByText(/Resultado:\s*\[\s*\]/)).toBeInTheDocument();
  });

  it('Debería calcular el resultado si no hay errores de validación', () => {
    (validateArrayNumbers as jest.Mock).mockReturnValue([]);
    (validateTarget as jest.Mock).mockReturnValue([]);
    (validateNumArray as jest.Mock).mockReturnValue([]);
    (TwoNumberSum as jest.Mock).mockReturnValue([2, 3]);
  
    render(<TowNumbers />);
  
    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: '1, 2, 3, 4' } });
    fireEvent.change(screen.getByPlaceholderText('Ej: 5'), { target: { value: '5' } });
    fireEvent.click(screen.getByText('Calcular'));
  
    expect(TwoNumberSum).toHaveBeenCalledWith([1, 2, 3, 4], 5);
    expect(screen.getByText(/Resultado:\s*\[\s*2, 3\s*\]/)).toBeInTheDocument();
  });

  it('Debería actualizar el estado de números al ingresar texto', () => {
    render(<TowNumbers />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: '1, 2, 3' } });

    expect(screen.getByPlaceholderText('Ej: 1, 2, 3, 4')).toHaveValue('1, 2, 3');
  });
  
});