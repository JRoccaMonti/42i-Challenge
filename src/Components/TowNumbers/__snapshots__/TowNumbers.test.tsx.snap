// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TowNumbers Component Debería coincidir con la snapshot 1`] = `
<DocumentFragment>
  <div
    class="towNumbersContainer"
  >
    <h3>
      Funcion Suma de dos números
    </h3>
    <div
      class="inputNumbersContainer"
    >
      <label>
        Ingrese el array de numeros (separados por comas):
        <input
          class="inputNumbers"
          placeholder="Ej: 1, 2, 3, 4"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      class="targetNumbersContainer"
    >
      <label>
        Ingrese la suma objetivo:
        <input
          class="inputNumbers"
          placeholder="Ej: 5"
          type="number"
          value=""
        />
      </label>
      <button>
        Calcular
      </button>
    </div>
    <div
      class="resultContainer"
    >
      <p>
        Resultado: [  ]
      </p>
    </div>
  </div>
</DocumentFragment>
`;
